pkg/ottl/ottlfuncs/func_formattime.go

```diff
@@ -0,0 +1,51 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type FormatTimeArguments[K any] struct {
+	Time   ottl.TimeGetter[K]
+	Format string
+}
+
+func NewFormatTimeFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("FormatTime", &FormatTimeArguments[K]{}, createFormatTimeFunction[K])
+}
+
+func createFormatTimeFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*FormatTimeArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("FormatTimeFactory args must be of type *FormatTimeArguments[K]")
+	}
+
+	return FormatTime(args.Time, args.Format)
+}
+
+func FormatTime[K any](timeValue ottl.TimeGetter[K], format string) (ottl.ExprFunc[K], error) {
+	if format == "" {
+		return nil, fmt.Errorf("format cannot be nil")
+	}
+
+	gotimeFormat, err := timeutils.StrptimeToGotime(format)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		t, err := timeValue.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
```

Comment:
We also need to validate if `t` is `nil` and return an error, otherwise the function would panic if that happens.

Reply:
looking at the implementation of [StandardTimeGetter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/expression.go#L845) this scenario cannot happen. First of all an error is always returned when the value is invalid and even if not, there is always an empty struct returned, which won't panic if the `Format()` function is called upon it

Reply:
Ops my bad, that's correct! sorry for the confusion.

Reply:
no worries, thanks for all other suggestions!